Pell's equation (#1388)
cp-algorithms/cp-algorithms: src/others/pells_equation.md

## To find the smallest positive solution
### Expressing the solution in terms of continued fractions
We can express the solution in terms of continued fractions. The continued fraction of $\sqrt{d}$ is periodic. Let's assume the continued fraction of $\sqrt{d}$ is $[a_{0}; \overline{a_{1}, a_{2}, \ldots, a_{r}}]$. The smallest positive solution is given by the convergent $[a_{0}; a_{1}, a_{2}, \ldots, a_{r}]$ where $r$ is the period of the continued fraction.

The convergents $p_{n}/q_{n}$ are the rational approximations to $\sqrt{d}$ obtained by truncating the continued fraction expansion at each stage. These convergents can be computed recursively.

Check whether the convergent satisfies the Pell's equation. If it does, then the convergent is the smallest positive solution.

Let's take an example to understand this by solving the equation $x^{2} - 2 \cdot y^{2} = 1$.
$\sqrt{2} = [1; \overline{2}] = 1 + 1/(2 + 1/(2 + 1/(2+ ...))) $. The convergents are $1/1, 3/2, 7/5, 17/12, 41/29, 99/70, \ldots$.
Now check for each convergent whether it satisfies the Pell's equation. The smallest positive solution is $3/2$.

### How to calculate the continued fraction of $\sqrt{d}$?

> Let's find the continued fraction of $\def\sf{\sqrt 7}\sf$.

Reviewer: The definition is broken in github previewer. Please just write \sqrt 7 unless you have a good reason to do a \define.

Reviewer 2: @jxu we don't really use GitHub previewer, as it breaks a lot of stuff (e.g. admonitions) and its rendering is inconsistent with CP-Algorithms anyway. Suggested approach is to use the [preview page](https://cp-algorithms.com/preview.html), or generate a preview on https://gh.cp-algorithms.com/1388/ (done automatically when you click on "Approve and run" for workflows).

---
tags:
  - Original
---

# Pell's Equation (Pell-Fermat Equation)

## Statement

> We are given a natural number $d$. We need to find the smallest positive integer $x$ such that $x^{2} - d \cdot y^{2} = 1$ for some integer $y$.

Reviewer 2: The smallest positive solution is actually $x=1$ and $y=0$. So, we want $x > 1$, not just positive.

Author: $ y = 0 $ is not positive

Reviewer 2: The paragraph currently says: smallest **positive** integer $x$ s.t. ... for **some** integer $y$.

It doesn't say that $y$ must be positive, only $x$.

## Solution
Here we will consider the case when $d$ is not a perfect square and $d>1$. The case when $d$ is a perfect square is trivial.
We can even assume that $d$ is square-free (i.e. it is not divisible by the square of any prime number) as we can absorb the factors of $d$ into $y$.

$x^{2} - d \cdot y^{2} = ( x + \sqrt{d} \cdot y ) ( x - \sqrt{d} \cdot y ) = 1$

The first part $( x + \sqrt{d} \cdot y )$ is always greater than 1. And the second part $( x - \sqrt{d} \cdot y )$ is always less than 1.

We will prove that all solutions to Pell's equation are given by powers of the smallest positive solution. Let's assume it to be 
$ x_{0} +  y_{0} \cdot \sqrt{d} $

We use method of descent to prove it. suppose there is a solution $u + v \cdot \sqrt{d}$ such that $u^{2} - d \cdot v^{2} = 1 $
Therefore, $ ( x_{0} + \sqrt{d} \cdot y_{0} )^{n}  < u + v \cdot \sqrt{d} < ( x_{0} + \sqrt{d} \cdot y_{0} )^{n+1} $

Multiplying the above inequality by $( x_{0} - \sqrt{d} \cdot y_{0} )^{n}$,(which is > 0 and < 1) we get

$ 1 < (u + v \cdot \sqrt{d})( x_{0} - \sqrt{d} \cdot y_{0} )^{n} < ( x_{0} + \sqrt{d} \cdot y_{0} ) $

> Because both $(u + v \cdot \sqrt{d})$ and $( x_{0} - \sqrt{d} \cdot y_{0} )^{n}$ have norm 1, their product is also a solution.

Reviewer 2: You use the term "norm" here, but it's not defined. We should add a definition + an explanation why we care about it (primarily because it is multiplicative).

> But this contradicts our assumption that $( x_{0} + \sqrt{d} \cdot y_{0} )$ is the smallest solution. Therefore, there is no solution between $( x_{0} + \sqrt{d} \cdot y_{0} )^{n}$ and $( x_{0} + \sqrt{d} \cdot y_{0} )^{n+1}$.

Reviewer 2: Please make sure that _everything_ renders correctly in the [preview](https://cp-algorithms.com/preview.html).

This looks like this for me:

![image](https://github.com/user-attachments/assets/aeb76d70-c864-4f72-9084-b85a03a8ae0b)

Author: Fixed all the markdown issues(except for one in nav bar)

> But this contradicts our assumption that $( x_{0} + \sqrt{d} \cdot y_{0} )$ is the smallest solution. Therefore, there is no solution between $( x_{0} + \sqrt{d} \cdot y_{0} )^{n}$ and $( x_{0} + \sqrt{d} \cdot y_{0} )^{n+1}$.

Reviewer 2: We defined the smallest solution by saying that it is the solution with the smallest positive $x$ (but we actually want $x > 1$). It also means that $x + y \sqrt d$ is the smallest possible, of course, but we should mention it explicitly abouve.

Hence, we conclude that all solutions are given by $( x_{0} + \sqrt{d} \cdot y_{0} )^{n}$ for some integer $n$.

> ## To find the smallest positive solution

Reviewer 2: ```suggestion
## Finding the smallest positive solution
```

### Expressing the solution in terms of continued fractions

> We can express the solution in terms of continued fractions. The continued fraction of $\sqrt{d}$ is periodic. Let's assume the continued fraction of $\sqrt{d}$ is $[a_{0}; \overline{a_{1}, a_{2}, \ldots, a_{r}}]$. The smallest positive solution is given by the convergent $[a_{0}; a_{1}, a_{2}, \ldots, a_{r}]$ where $r$ is the period of the continued fraction.

Reviewer 2: Let's add a link to [continued fractions article](https://cp-algorithms.com/algebra/continued-fractions.html). Ideally, I would also add explanation on why it is a solution at all, and why is it the smallest possible.

Reviewer 2: Also, the paragraph says that "the smallest solution is given by ...", but doesn't tell that, specifically, the solution is $p_n^2 - q_n^2 d = 1$. Instead, you just write below that $p_n / q_n$ is the solution.

The convergents $p_{n}/q_{n}$ are the rational approximations to $\sqrt{d}$ obtained by truncating the continued fraction expansion at each stage. These convergents can be computed recursively.

Check whether the convergent satisfies the Pell's equation. If it does, then the convergent is the smallest positive solution.

Let's take an example to understand this by solving the equation $x^{2} - 2 \cdot y^{2} = 1$.

> $\sqrt{2} = [1; \overline{2}] = 1 + 1/(2 + 1/(2 + 1/(2+ ...))) $. The convergents are $1/1, 3/2, 7/5, 17/12, 41/29, 99/70, \ldots$.

Reviewer 2: LaTeX is misrendered here too:

![image](https://github.com/user-attachments/assets/02607d95-a14e-4a1f-9480-05fa67c48956)

Author: Fixed

Now check for each convergent whether it satisfies the Pell's equation. The smallest positive solution is $3/2$.

### How to calculate the continued fraction of $\sqrt{d}$?
Let's find the continued fraction of $\def\sf{\sqrt 7}\sf$.

$\sf \approx 2.6457 = 2 + 0.6457$

> $a_{0} = 2 $

Reviewer 2: This misrenders too because of the space near `$`.

Author: Fixed

Subtract $a_{0}$ from the number and take the reciprocal of the remaining number.

That is, we calculate ${1\over \sf - 2} \approx 1.5486$. The integer part $a_{1}$ is $1$.
So:

>     $$\sf=2+\cfrac{1}{1+\cfrac1{\vdots}}$$ Where we haven't calculated the $\vdots$ part yet.

Reviewer 2: This misrenders too. You need to put newlines before and after each `$$...$$` block for it to render correctly.

Author: Fixed

To get that, we subtract $a_{1}$ from the number and take the reciprocal of the remaining number. That is, we calculate ${1\over 1.5486 - 1} \approx 1.8228$. The integer part $a_{2}$ is $1$.
So:
    $$\sf=2+\cfrac{1}{1+\cfrac1{1+\cfrac1{\vdots}}}$$
Now ${1\over 1.8228 - 1} \approx 1.2153$. So $a_{3} = 1$.
Continuing this process, ${1\over 1.2153 - 1} \approx 4.645$. So $a_{4} = 4$.

$$\sf=2+\cfrac{1}{1+\cfrac1{1+\cfrac1{1+\cfrac4{\vdots}}}}$$

we get the continued fraction of $\sf$ as $[2; 1, 1, 4, 1, 1, 4, \ldots]$.

It can also be calculated using only [integer based calculation](https://cp-algorithms.com/algebra/continued-fractions.html) 

> ### Finding the solution using Chakravala method

Reviewer 2: This looks complicated. Are there specific reasons to do it over continued fractions variant? I find the exposition very hard to read (largely because of misrenders, though).

Unless it's in some way better than continued fractions method, I'd consider dropping the section altogether.

Author: Added link to continued fractions and removed floating point based

Reviewer 2: Thanks! But I'm still not convinced this section is needed at all. It seems much more complicated than doing it with convergents. I really think we should either have a good reason to include it (e.g. mention why is it better than convergents), or remove it.

The Chakravala method is an ancient Indian algorithm to solve Pell's equation. It is based on the Brahmagupta's identity of quadratic decomposition 
$(x_{1}^{2} - n \cdot y_{1}^{2}) \cdot (x_{2}^{2} - n \cdot y_{2}^{2}) = (x_{1} \cdot x_{2} + n \cdot y_{1} \cdot y_{2})^{2} - n \cdot (x_{1} \cdot y_{2} + x_{2} \cdot y_{1})^{2}$
$(x_{1}^{2} - n \cdot y_{1}^{2}) \cdot (x_{2}^{2} - n \cdot y_{2}^{2}) = (x_{1} \cdot x_{2} - n \cdot y_{1} \cdot y_{2})^{2} - n \cdot (x_{1} \cdot y_{2} - x_{2} \cdot y_{1})^{2}$

> And Bhaskara's Lemma:

Reviewer 2: This, and some further sections below still misrender in preview. Please fix all of them.

            // Adjust m to ensure m < sqrt(n) < m + k
            while (m > sqrt(n)) m -= absK;
            while (m + absK < sqrt_n) m += absK;

            // Select closest value to n
            if (abs(m * m - n) > abs((m + absK) * (m + absK) - n)) {
                m = m + absK;
            }
        }

        // Print the current triple
        cout << "[a = " << a << ", b = " << b << ", k = " << k << "]" << endl;

        // Update a, b, k using the recurrence relations
        long long alpha = a;
        a = (m * a + n * b) / absK;
        b = (alpha + m * b) / absK;
        k = (m * m - n) / k;

        // Increment step counter
        steps++;
    }

    // Print final result
    cout << a << "^2 - " << n << " x " << b << "^2 = 1 in " << steps << " calculations." << endl;

    // Return the solution as a pair (a, b)
    return {a, b};
}

```

## References
- [Pell's equation - Wikipedia](https://en.wikipedia.org/wiki/Pell%27s_equation)
- [Periodic Continued Fractions](https://en.wikipedia.org/wiki/Periodic_continued_fraction)
- [Chakravala Method](http://publications.azimpremjifoundation.org/1630/1/3_The%20Chakravala%20Method.pdf)

> - [Codeforces Pell's Equation](https://codeforces.com/topic/116937/en20)

Reviewer 2: ```suggestion
- [Pythagorean triples and Pell's equations - Codeforces](https://codeforces.com/blog/entry/116313)
```

This factorization is important because it shows the connection to quadratic irrationals. The first part $(x + y \sqrt{d})$ is always greater than 1, and the second part $(x - y \sqrt{d})$ is always less than 1, since their product is 1.

The norm of an expression $u + v \sqrt{d}$ is defined as $N(u + v \sqrt{d}) = (u + v \sqrt{d})(u - v \sqrt{d}) = u^2 - d v^2$. The norm is multiplicative: $N(ab) = N(a)N(b)$. This property is crucial in the descent argument below.

We will prove that all solutions to Pell's equation are given by powers of the smallest positive solution. Let's assume it to be $x_0 + y_0 \sqrt{d}$, where $x_0 > 1$ is the smallest possible value for $x$.

## Method of Descent
Suppose there is a solution $u + v \cdot \sqrt{d}$ such that $u^{2} - d \cdot v^{2} = 1$ and is not a power of  $( x_{0} + \sqrt{d} \cdot y_{0} )$
Then it must lie between two powers of $( x_{0} + \sqrt{d} \cdot y_{0} )$. 
i.e, For some n, $( x_{0} + \sqrt{d} \cdot y_{0} )^{n}  < u + v \cdot \sqrt{d} < ( x_{0} + \sqrt{d} \cdot y_{0} )^{n+1}$

Multiplying the above inequality by $( x_{0} - \sqrt{d} \cdot y_{0} )^{n}$,(which is > 0 and < 1) we get

> $1 < (u + v \cdot \sqrt{d})( x_{0} - \sqrt{d} \cdot y_{0} )^{n} < ( x_{0} + \sqrt{d} \cdot y_{0} )$

Reviewer 2: ```suggestion
$$
1 < (u + v \cdot \sqrt{d})( x_{0} - \sqrt{d} \cdot y_{0} )^{n} < ( x_{0} + \sqrt{d} \cdot y_{0} )
$$

```

---
tags:
  - Original
---

# Pell's Equation (Pell-Fermat Equation)

## Statement
We are given a natural number $d$. We need to find the smallest positive integer $x$ such that $x^{2} - d \cdot y^{2} = 1$ for some integer $y$.

> OR
> 
> We want to find all the possible solutions of the equation $x^{2} - d \cdot y^{2} = 1$.

Reviewer 2: ```suggestion
Alternative formulation: We want to find all the possible solutions of the equation $x^{2} - d \cdot y^{2} = 1$.
```

## Solution

> Here we will consider the case when $d$ is not a perfect square and $d>1$. The case when $n$ is a perfect square is trivial.

Reviewer 2: ```suggestion
Here we will consider the case when $d$ is not a perfect square and $d>1$. The case when $d$ is a perfect square is trivial.
```

> We can even assume that $d$ is square-free (i.e. it is not divisible by the square of any prime number) as we can absorb the factors of $d$ into y.

Reviewer 2: ```suggestion
We can even assume that $d$ is square-free (i.e. it is not divisible by the square of any prime number) as we can absorb the factors of $d$ into $y$.
```

The first part $( x + \sqrt{d} \cdot y )$ is always greater than 1. And the second part $( x - \sqrt{d} \cdot y )$ is always less than 1.

We will prove that all solutions to Pell's equation are given by powers of the smallest positive solution. Let's assume it to be 
$ x_{0} +  y_{0} \cdot \sqrt{d} $

[//]: # (We claim that every solution to the equation is given by $ &#40;x_{0} +  y_{0} \cdot \sqrt{d}&#41;^{n} $ for some integer $ n $. The idea is that any other solution)
[//]: # (  $ &#40; x + \sqrt{d} \cdot y &#41;$ must be of this form, meaning that all solutions are generated by repeated multiplication of the smallest solution. Here we use Brahmagupta's identity of quadratic decomposition.)
[//]: # (   $ &#40;a^{2} - n \cdot b^{2}&#41; \cdot &#40;c^{2} - n \cdot d^{2}&#41; = &#40;a \cdot c + n \cdot b \cdot d&#41;^{2} - n \cdot &#40;a \cdot d + b \cdot c&#41;^{2}$)
[//]: # (   )
[//]: # (Here, $a = x_{1}$, $b = y_{1}$, $c = x_{2}$, $d = y_{2}$)
[//]: # (   So if &#40;x_{1}, y_{1}&#41; and &#40;x_{2}, y_{2}&#41; are solutions, then $&#40;x_{1} \cdot x_{2} + n \cdot y_{1} \cdot y_{2}, x_{1} \cdot y_{2} + y_{1} \cdot x_{2}&#41;$ is also a solution.)

> [//]: # (First, we prove that product of two solutions is also a solution.)

Reviewer 2: Why does it use `&#40;` and `&#41;`? It doesn't render properly at https://gh.cp-algorithms.com/1388/others/pells_equation.html.

Reviewer 2: Generally, please check your suggestion at https://cp-algorithms.com/preview.html, as many things don't render properly at the actual website.

Author: `[//]: ` means comment in markdown. removed them

            while (m > sqrt(n)) m -= absK;
            while (m + absK < sqrt_n) m += absK;

            // Select closest value to n
            if (abs(m * m - n) > abs((m + absK) * (m + absK) - n)) {
                m = m + absK;
            }
        }

        // Print the current triple
        cout << "[a = " << a << ", b = " << b << ", k = " << k << "]" << endl;

        // Update a, b, k using the recurrence relations
        long long alpha = a;
        a = (m * a + n * b) / absK;
        b = (alpha + m * b) / absK;
        k = (m * m - n) / k;

        // Increment step counter
        steps++;
    }

    // Print final result
    cout << a << "^2 - " << n << " x " << b << "^2 = 1 in " << steps << " calculations." << endl;

    // Return the solution as a pair (a, b)
    return {a, b};
}

```

## References
- [Pell's equation - Wikipedia](https://en.wikipedia.org/wiki/Pell%27s_equation)
- [Periodic Continued Fractions](https://en.wikipedia.org/wiki/Periodic_continued_fraction)
- [Chakravala Method - Wikipedia](https://en.wikipedia.org/wiki/Chakravala_method)
- [Chakravala Method](http://publications.azimpremjifoundation.org/1630/1/3_The%20Chakravala%20Method.pdf)

> - [Chakravala Method](https://www.isibang.ac.in/~sury/chakravala.pdf)

Reviewer 2: Do we really need all 3 here? Maybe picking one of them is okay?

Author: removed 1st and 3rd

$\sqrt{2} = [1; \overline{2}] = 1 + 1/(2 + 1/(2 + 1/(2+ ...))) $. The convergents are $1/1, 3/2, 7/5, 17/12, 41/29, 99/70, \ldots$.
Now check for each convergent whether it satisfies the Pell's equation. The smallest positive solution is $3/2$.

### How to calculate the continued fraction of $\sqrt{d}$?
Let's find the continued fraction of $\def\sf{\sqrt 7}\sf$.

$\sf \approx 2.6457 = 2 + 0.6457$

$a_{0} = 2 $

Subtract $a_{0}$ from the number and take the reciprocal of the remaining number.

That is, we calculate ${1\over \sf - 2} \approx 1.5486$. The integer part $a_{1}$ is $1$.
So:
    $$\sf=2+\cfrac{1}{1+\cfrac1{\vdots}}$$ Where we haven't calculated the $\vdots$ part yet.
To get that, we subtract $a_{1}$ from the number and take the reciprocal of the remaining number. That is, we calculate ${1\over 1.5486 - 1} \approx 1.8228$. The integer part $a_{2}$ is $1$.
So:
    $$\sf=2+\cfrac{1}{1+\cfrac1{1+\cfrac1{\vdots}}}$$
Now ${1\over 1.8228 - 1} \approx 1.2153$. So $a_{3} = 1$.
Continuing this process, ${1\over 1.2153 - 1} \approx 4.645$. So $a_{4} = 4$.

$$\sf=2+\cfrac{1}{1+\cfrac1{1+\cfrac1{1+\cfrac4{\vdots}}}}$$

> we get the continued fraction of $\sf$ as $[2; 1, 1, 4, 1, 1, 4, \ldots]$.

Reviewer: Note this relies on floating point numbers. You can do it entirely with integer arithmetic.

Author: Can you give me reference for doing it using integer arithmetic completely?

Reviewer 2: "Quadratic irrationality" section of [continued fraction article](https://cp-algorithms.com/algebra/continued-fractions.html) has it:

```py
# compute the continued fraction of sqrt(n)
def sqrt(n):
    n0 = math.floor(math.sqrt(n))
    x, y, z = 1, 0, 1
    a = []
    def step(x, y, z):
        a.append((x * n0 + y) // z)
        t = y - a[-1]*z
        x, y, z = -z*x, z*t, t**2 - n*x**2
        g = math.gcd(x, math.gcd(y, z))
        return x // g, y // g, z // g

    used = dict()
    for i in range(n):
        used[x, y, z] = i
        x, y, z = step(x, y, z)
        if (x, y, z) in used:
            return a
```

Author: included as reference

Reviewer 2: I would actually strongly suggest to replace the whole section to only show integer calculation, as floating-point calculation might be unstable.
